# Проект 1. Кто хочет стать миллионером?

В этом модуле вы будете проводить предварительный анализ данных на основе датасета сайта IMBD.

In [52]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
from collections import Counter

In [53]:
data = pd.read_csv('../data/imbd-sf.csv')
data.head(5)

imdb_id  popularity     budget     revenue                original_title  \
0  tt0369610   32.985763  150000000  1513528810                Jurassic World   
1  tt1392190   28.419936  150000000   378436354            Mad Max: Fury Road   
2  tt2908446   13.112507  110000000   295238201                     Insurgent   
3  tt2488496   11.173104  200000000  2068178225  Star Wars: The Force Awakens   
4  tt2820852    9.335014  190000000  1506249360                     Furious 7   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   

                         tagline  \
0              The park is open.   
1             What a Lovely Day.   
2     One Choice Can Destroy You   
3  Every generation has a story.   
4            Vengeance Hits Home   

                                            overview  runtime  \
0  Twenty-two years after the events of Jurassic ...      124   
1  An apocalyptic story set in the furthest reach...      120   
2  Beatrice Prior must confront her inner demons ...      119   
3  Thirty years after defeating the Galactic Empi...      136   
4  Deckard Shaw seeks revenge against Dominic Tor...      137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date  vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015        5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015        6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015        2480   
3          Lucasfilm|Truenorth Productions|Bad Robot   12/15/2015        5292   
4  Universal Pictures|Original Film|Media Rights ...     4/1/2015        2947   

   vote_average  release_year  
0           6.5          2015  
1           7.1          2015  
2           6.3          2015  
3           7.5          2015  
4           7.3          2015

In [54]:
len(data)

1890

# Предобработка датасета

In [98]:
# Прибыль — разность между сборами и бюджетом фильма.
data['income'] = data['revenue'] - data['budget']

# Add release month.
data['release_month'] = data['release_date'].apply(lambda date: date.split('/')[0])

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. **The Warrior's Way	(tt1032751)**
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [56]:
movie = data.loc[data['budget'] == data['budget'].max()]
display(movie)

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     income  
491    12/2/2010          74           6.4          2010 -413912431

# 2. Какой из фильмов самый длительный (в минутах)
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. **Gods and Generals	(tt0279111)**
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [57]:
movie = data.loc[data['runtime'] == data['runtime'].max()]
display(movie)

imdb_id  popularity    budget   revenue     original_title  \
1158  tt0279111    0.469518  56000000  12923936  Gods and Generals   

                                                   cast           director  \
1158  Stephen Lang|Jeff Daniels|Robert Duvall|Kevin ...  Ronald F. Maxwell   

                                  tagline  \
1158  The nations heart was touched by...   

                                               overview  runtime  \
1158  The film centers mostly around the personal an...      214   

                 genres                production_companies release_date  \
1158  Drama|History|War  Turner Pictures|Antietam Filmworks    2/21/2003   

      vote_count  vote_average  release_year    income  
1158          23           5.8          2003 -43076064

# 3. Какой из фильмов самый короткий (в минутах)
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. **Winnie the Pooh	tt1449283**
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [58]:
movie = data.loc[data['runtime'] == data['runtime'].min()]
display(movie)

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  \
769  During an ordinary day in Hundred Acre Wood, W...       63   

               genres                               production_companies  \
769  Animation|Family  Walt Disney Pictures|Walt Disney Animation Stu...   

    release_date  vote_count  vote_average  release_year    income  
769    4/13/2011         174           6.8          2011 -15540000

# 4. Какое число ближе к средней длительности фильма в датасете?

Варианты ответов:
1. 115
2. **110**
3. 105
4. 120
5. 100


In [59]:
runtime = data['runtime'].mean()
display(runtime)

109.65343915343915

# 5. Какое число ближе к медианной длительности фильма в датасете?
Варианты ответов:
1. **106**
2. 112
3. 101
4. 120
5. 115




In [60]:
runtime = data['runtime'].median()
display(runtime)

106.5

# 6. Какой самый прибыльный фильм?

Внимание! Здесь и далее под «прибылью» или «убытками» понимается разность между сборами и бюджетом фильма.

Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. **Avatar	tt0499549**

In [61]:
# sorted_data = data.sort_values(by='income', ascending=False)
# display(sorted_data)

movie = data.loc[data['income'] == data['income'].max()]
display(movie)

imdb_id  popularity     budget     revenue original_title  \
239  tt0499549    9.432768  237000000  2781505847         Avatar   

                                                  cast       director  \
239  Sam Worthington|Zoe Saldana|Sigourney Weaver|S...  James Cameron   

                         tagline  \
239  Enter the World of Pandora.   

                                              overview  runtime  \
239  In the 22nd century, a paraplegic Marine is di...      162   

                                       genres  \
239  Action|Adventure|Fantasy|Science Fiction   

                                  production_companies release_date  \
239  Ingenious Film Partners|Twentieth Century Fox ...   12/10/2009   

     vote_count  vote_average  release_year      income  
239        8458           7.1          2009  2544505847

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. **The Warrior's Way tt1032751**
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [62]:
movie = data.loc[data['income'] == data['income'].min()]
display(movie)

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     income  
491    12/2/2010          74           6.4          2010 -413912431

# 8. У скольких фильмов из датасета объем сборов оказался выше бюджета?
Варианты ответов:
1. **1478**
2. 1520
3. 1241
4. 1135
5. 1398


In [63]:
display(len(data.loc[data['revenue'] > data['budget']]))

1478

# 9. Какой фильм оказался самым кассовым в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. **The Dark Knight	tt0468569**
5. Mamma Mia!	tt0795421

In [64]:
movies = data.loc[data['release_year'] == 2008]
# display(movies)

movie = movies.loc[movies['revenue'] == movies['revenue'].max()]
display(movie)

imdb_id  popularity     budget     revenue   original_title  \
600  tt0468569    8.466668  185000000  1001921825  The Dark Knight   

                                                  cast           director  \
600  Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   

             tagline                                           overview  \
600  Why So Serious?  Batman raises the stakes in his war on crime. ...   

     runtime                       genres  \
600      152  Drama|Action|Crime|Thriller   

                                  production_companies release_date  \
600  DC Comics|Legendary Pictures|Warner Bros.|Syncopy    7/16/2008   

     vote_count  vote_average  release_year     income  
600        8432           8.1          2008  816921825

# 10. Самый убыточный фильм за период с 2012 по 2014 (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. **The Lone Ranger	tt1210819**


In [65]:
movies = data.loc[(data['release_year'] >= 2012) & (data['release_year'] <= 2014)]
# display(movies)

movie = movies.loc[movies['income'] == movies['income'].min()]
display(movie)

imdb_id  popularity     budget   revenue   original_title  \
1246  tt1210819     1.21451  255000000  89289910  The Lone Ranger   

                                                   cast        director  \
1246  Johnny Depp|Armie Hammer|William Fichtner|Hele...  Gore Verbinski   

                      tagline  \
1246  Never Take Off the Mask   

                                               overview  runtime  \
1246  The Texas Rangers chase down a gang of outlaws...      149   

                        genres  \
1246  Action|Adventure|Western   

                                   production_companies release_date  \
1246  Walt Disney Pictures|Jerry Bruckheimer Films|I...     7/3/2013   

      vote_count  vote_average  release_year     income  
1246        1607           6.0          2013 -165710090

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. **Drama**
4. Comedy
5. Thriller

In [66]:
genres = Counter()
for genre in ['Action', 'Adventure', 'Drama', 'Comedy', 'Thriller']:
    genres[genre] = len(data.loc[data['genres'].str.contains(genre)])
    
display(genres)

genres = pd.Series(genres, index=['Action', 'Adventure', 'Drama', 'Comedy', 'Thriller'])
display(genres.idxmax())

Counter({'Action': 583,
         'Adventure': 416,
         'Drama': 782,
         'Comedy': 683,
         'Thriller': 597})

'Drama'

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. **Drama**
2. Comedy
3. Action
4. Thriller
5. Adventure

In [67]:
genres = Counter()
for genre in ['Action', 'Adventure', 'Drama', 'Comedy', 'Thriller']:
    genres[genre] = len(data.loc[(data['genres'].str.contains(genre)) & (data['income'] > 0)])
    
display(genres)

genres = pd.Series(genres, index=['Action', 'Adventure', 'Drama', 'Comedy', 'Thriller'])
display(genres.idxmax())

Counter({'Action': 444,
         'Adventure': 337,
         'Drama': 560,
         'Comedy': 551,
         'Thriller': 446})

'Drama'

# 13. Кто из режиссеров снял больше всего фильмов?
Варианты ответов:
1. Steven Spielberg
2. Ridley Scott 
3. **Steven Soderbergh**
4. Christopher Nolan
5. Clint Eastwood

In [68]:
directors_list = ['Steven Spielberg', 'Ridley Scott', 'Steven Soderbergh', 'Christopher Nolan', 'Clint Eastwood']

directors = Counter()
for director in directors_list:
    directors[director] = len(data.loc[data['director'].str.match(director)])
    
display(directors)

directors = pd.Series(directors, index=directors_list)
display(directors.idxmax())

Counter({'Steven Spielberg': 10,
         'Ridley Scott': 12,
         'Steven Soderbergh': 13,
         'Christopher Nolan': 8,
         'Clint Eastwood': 12})

'Steven Soderbergh'

# 14. Кто из режиссеров снял больше всего Прибыльных фильмов?
Варианты ответов:
1. Steven Soderbergh
2. Clint Eastwood
3. Steven Spielberg
4. **Ridley Scott**
5. Christopher Nolan

In [69]:
directors_list = ['Steven Spielberg', 'Ridley Scott', 'Steven Soderbergh', 'Christopher Nolan', 'Clint Eastwood']

directors = Counter()
for director in directors_list:
    directors[director] = len(data.loc[(data['director'].str.match(director)) & (data['income'] > 0)])
    
display(directors)

directors = pd.Series(directors, index=directors_list)
display(directors.idxmax())

Counter({'Steven Spielberg': 10,
         'Ridley Scott': 12,
         'Steven Soderbergh': 10,
         'Christopher Nolan': 8,
         'Clint Eastwood': 10})

'Ridley Scott'

# 15. Кто из режиссеров принес больше всего прибыли?
Варианты ответов:
1. Steven Spielberg
2. Christopher Nolan
3. David Yates
4. James Cameron
5. **Peter Jackson**


In [70]:
directors_list = ['Steven Spielberg', 'David Yates', 'James Cameron', 'Christopher Nolan', 'Peter Jackson']

directors = Counter()
for director in directors_list:
    directors[director] = data.loc[data['director'].str.match(director)]['income'].sum()
    
display(directors)

directors = pd.Series(directors, index=directors_list)
display(directors.idxmax())

Counter({'Steven Spielberg': 2449700791,
         'David Yates': 3379295625,
         'James Cameron': 2548546718,
         'Christopher Nolan': 3162548502,
         'Peter Jackson': 5202593685})

'Peter Jackson'

# 16. Какой актер принес больше всего прибыли?
Варианты ответов:
1. **Emma Watson**
2. Johnny Depp
3. Michelle Rodriguez
4. Orlando Bloom
5. Rupert Grint

In [71]:
artists_list = ['Emma Watson', 'Johnny Depp', 'Michelle Rodriguez', 'Orlando Bloom', 'Rupert Grint']

artists = Counter()
for artist in artists_list:
    artists[artist] = data.loc[data['cast'].str.contains(artist)]['income'].sum()
    
display(artists)

artists = pd.Series(artists, index=artists_list)
display(artists.idxmax())

Counter({'Emma Watson': 6666245597,
         'Johnny Depp': 4776417000,
         'Michelle Rodriguez': 4608031235,
         'Orlando Bloom': 5148578162,
         'Rupert Grint': 6408638290})

'Emma Watson'

# 17. Какой актер принес меньше всего прибыли в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Danny Huston
3. **Kirsten Dunst**
4. Jim Sturgess
5. Sami Gayle

In [72]:
artists_list = ['Nicolas Cage', 'Danny Huston', 'Kirsten Dunst', 'Jim Sturgess', 'Sami Gayle']

artists = Counter()
for artist in artists_list:
    artists[artist] = data.loc[(data['release_year'] == 2012) & (data['cast'].str.contains(artist))]['income'].sum()
    
display(artists)

artists = pd.Series(artists, index=artists_list)
display(artists.idxmin())

Counter({'Nicolas Cage': -32893443,
         'Danny Huston': -32893443,
         'Kirsten Dunst': -68109207,
         'Jim Sturgess': -23410657,
         'Sami Gayle': -32893443})

'Kirsten Dunst'

# 18. Какой актер снялся в большем количестве высокобюджетных фильмов? Примечание: в фильмах, где бюджет выше среднего по данной выборке.
Варианты ответов:
1. Tom Cruise
2. Mark Wahlberg 
3. **Matt Damon**
4. Angelina Jolie
5. Adam Sandler

In [73]:
artists_list = ['Tom Cruise', 'Mark Wahlberg', 'Matt Damon', 'Angelina Jolie', 'Adam Sandler']

artists = Counter()
for artist in artists_list:
    artists[artist] = len(data.loc[(data['cast'].str.contains(artist)) & (data['budget'] > data['budget'].mean())])
    
display(artists)

artists = pd.Series(artists, index=artists_list)
display(artists.idxmax())

Counter({'Tom Cruise': 15,
         'Mark Wahlberg': 14,
         'Matt Damon': 18,
         'Angelina Jolie': 16,
         'Adam Sandler': 17})

'Matt Damon'

# 19. В фильмах какого жанра больше всего снимался Nicolas Cage?  
Варианты ответа:
1. Drama
2. **Action**
3. Thriller
4. Adventure
5. Crime

In [74]:
genres_list = ['Drama', 'Action', 'Thriller', 'Adventure', 'Crime']

genres = Counter()
for genre in genres_list:
    genres[genre] = len(data.loc[(data['genres'].str.contains(genre)) & (data['cast'].str.contains('Nicolas Cage'))])
    
display(genres)

genres = pd.Series(genres, index=genres_list)
display(genres.idxmax())

Counter({'Drama': 12,
         'Action': 17,
         'Thriller': 15,
         'Adventure': 7,
         'Crime': 10})

'Action'

# 20. Какая студия сняла больше всего фильмов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [75]:
production_companies_list = ['Universal Pictures (Universal)', 'Paramount Pictures', 'Columbia Pictures', 'Warner Bros', 'Twentieth Century Fox Film Corporation']

production_companies = Counter()
for production_company in production_companies_list:
    production_companies[production_company] = len(data.loc[data['production_companies'].str.contains(production_company)])
#     production_companies[production_company] = len(data.loc[data['production_companies'].str.match(production_company)])
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

universal = data.loc[data['production_companies'].str.contains('Universal Pictures')]
display(universal)

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/strings.py:1952: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


Counter({'Universal Pictures (Universal)': 0,
         'Paramount Pictures': 122,
         'Columbia Pictures': 163,
         'Warner Bros': 183,
         'Twentieth Century Fox Film Corporation': 109})

'Warner Bros'

imdb_id  popularity     budget     revenue  \
4     tt2820852    9.335014  190000000  1506249360   
8     tt2293640    7.404165   74000000  1156730962   
12    tt0470752    6.118847   15000000    36869414   
26    tt2637276    4.564549   68000000   215863606   
31    tt2848292    3.877764   29000000   287506194   
...         ...         ...        ...         ...   
1839  tt0218967    0.911808   60000000   124745083   
1841  tt0192614    0.848261   15000000    35007180   
1855  tt0158622    0.632783   83000000    59468275   
1871  tt0144528    1.616724   84000000   123307945   
1887  tt0131704    0.104908   76000000    35134820   

                            original_title  \
4                                Furious 7   
8                                  Minions   
12                              Ex Machina   
26                                   Ted 2   
31                         Pitch Perfect 2   
...                                    ...   
1839                        The Family Man   
1841                            The Skulls   
1855    The Flintstones in Viva Rock Vegas   
1871        Nutty Professor II: The Klumps   
1887  The Adventures of Rocky & Bullwinkle   

                                                   cast  \
4     Vin Diesel|Paul Walker|Jason Statham|Michelle ...   
8     Sandra Bullock|Jon Hamm|Michael Keaton|Allison...   
12    Domhnall Gleeson|Alicia Vikander|Oscar Isaac|S...   
26    Mark Wahlberg|Seth MacFarlane|Amanda Seyfried|...   
31    Anna Kendrick|Rebel Wilson|Hailee Steinfeld|Br...   
...                                                 ...   
1839  Nicolas Cage|Téa Leoni|Don Cheadle|Saul Rubin...   
1841  Joshua Jackson|Paul Walker|Hill Harper|Leslie ...   
1855  Mark Addy|Stephen Baldwin|Kristen Johnston|Jan...   
1871  Eddie Murphy|Janet Jackson|Larry Miller|John A...   
1887  Rene Russo|Jason Alexander|Piper Perabo|Randy ...   

                      director  \
4                    James Wan   
8     Kyle Balda|Pierre Coffin   
12                Alex Garland   
26             Seth MacFarlane   
31             Elizabeth Banks   
...                        ...   
1839              Brett Ratner   
1841                 Rob Cohen   
1855              Brian Levant   
1871               Peter Segal   
1887               Des McAnuff   

                                                tagline  \
4                                   Vengeance Hits Home   
8          Before Gru, they had a history of bad bosses   
12    There is nothing more human than the will to s...   
26                                Ted is Coming, Again.   
31                                   We're back pitches   
...                                                 ...   
1839  What if you made different choices? What if yo...   
1841       Getting in is easy. Getting out is a killer.   
1855                                 Get ready to rock!   
1871                               The Klumps are back!   
1887            This summer it's not the same old bull.   

                                               overview  runtime  \
4     Deckard Shaw seeks revenge against Dominic Tor...      137   
8     Minions Stuart, Kevin and Bob are recruited by...       91   
12    Caleb, a 26 year old coder at the world's larg...      108   
26    Newlywed couple Ted and Tami-Lynn want to have...      115   
31    The Bellas are back, and they are better than ...      115   
...                                                 ...      ...   
1839  Jack's lavish, fast-paced lifestyle changes on...      125   
1841  Luke's exultance at being selected for The Sku...      106   
1855  The Flintstones are at it again. The Flintston...       90   
1871  The hilarity begins when professor Sherman Klu...      106   
1887  Rocky and Bullwinkle have been living off the ...       88   

                                        genres  \
4                        Action|Crime|Thriller   
8            Family|Animation|Adventure|Comedy   
12          

# 21. Какая студия сняла больше всего фильмов в 2015 году?
Варианты ответа:
1. Universal Pictures
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [76]:
production_companies_list = ['Universal Pictures', 'Paramount Pictures', 'Columbia Pictures', 'Warner Bros', 'Twentieth Century Fox Film Corporation']

production_companies = Counter()
for production_company in production_companies_list:
#     production_companies[production_company] = len(data.loc[(data['release_year'] == 2015) & (data['production_companies'].str.contains(production_company))])
    production_companies[production_company] = len(data.loc[(data['release_year'] == 2015) & (data['production_companies'].str.match(production_company))])
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

Counter({'Universal Pictures': 9,
         'Paramount Pictures': 7,
         'Columbia Pictures': 7,
         'Warner Bros': 3,
         'Twentieth Century Fox Film Corporation': 8})

'Universal Pictures'

# 22. Какая студия заработала больше всего денег в жанре комедий за все время?
Варианты ответа:
1. Warner Bros
2. **Universal Pictures (Universal)**
3. Columbia Pictures
4. Paramount Pictures
5. Walt Disney

In [77]:
production_companies_list = ['Universal Pictures', 'Paramount Pictures', 'Columbia Pictures', 'Warner Bros', 'Walt Disney']

production_companies = Counter()
for production_company in production_companies_list:
    production_companies[production_company] = data.loc[(data['genres'].str.match('Comedy')) & (data['production_companies'].str.contains(production_company))]['income'].sum()
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

Counter({'Universal Pictures': 5121151527,
         'Paramount Pictures': 1905015448,
         'Columbia Pictures': 2956105621,
         'Warner Bros': 2430748470,
         'Walt Disney': 1874968657})

'Universal Pictures'

# 23. Какая студия заработала больше всего денег в 2012 году?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. **Columbia Pictures**
4. Paramount Pictures
5. Lucasfilm

In [93]:
production_companies_list = ['Universal Pictures', 'Paramount Pictures', 'Columbia Pictures', 'Warner Bros', 'Walt Disney']

production_companies = Counter()
for production_company in production_companies_list:
    production_companies[production_company] = data.loc[(data['release_year'] == 2012) & (data['production_companies'].str.contains(production_company))]['income'].sum()
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

Counter({'Universal Pictures': 1981011579,
         'Paramount Pictures': 410726715,
         'Columbia Pictures': 2501406608,
         'Warner Bros': 1258020056,
         'Walt Disney': 726836264})

'Columbia Pictures'

# 24. Самый убыточный фильм от Paramount Pictures?
Варианты ответа:

1. **K-19: The Widowmaker tt0267626**
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [94]:
paramount_movies = data.loc[(data['production_companies'].str.contains('Paramount Pictures'))]
movie = paramount_movies.loc[paramount_movies['income'] == paramount_movies['income'].min()]
display(movie)

imdb_id  popularity     budget   revenue        original_title  \
926  tt0267626     0.72233  100000000  35168966  K-19: The Widowmaker   

                                                  cast         director  \
926  Harrison Ford|Liam Neeson|Peter Sarsgaard|Joss...  Kathryn Bigelow   

                      tagline  \
926  Fate has found its hero.   

                                              overview  runtime  ...  \
926  When Russia's first nuclear submarine malfunct...      138  ...   

                                  production_companies release_date  \
926  Paramount Pictures|Intermedia Films|National G...    7/19/2002   

    vote_count  vote_average  release_year    income  release_month  \
926        146           6.0          2002 -64831034              7   

    movie_title_chars_count  movie_title_words_count         movie_title_words  
926                      20                        3  [k-19:, the, widowmaker]  

[1 rows x 21 columns]

# 25. Какой Самый прибыльный год (заработали больше всего)?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. **2015**

In [96]:
years_list = [2014, 2008, 2012, 2002, 2015]

years = Counter()
for year in years_list:
    years[year] = data.loc[(data['release_year'] == year)]['income'].sum()
    
display(years)

years = pd.Series(years, index=years_list)
display(years.idxmax())

Counter({2014: 16397812953,
         2008: 11663881990,
         2012: 16077001687,
         2002: 9002361487,
         2015: 18668572378})

2015

# 26. Какой Самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. **2014**
2. 2008
3. 2012
4. 2010
5. 2015

In [81]:
years_list = [2014, 2008, 2012, 2010, 2015]

years = Counter()
for year in years_list:
    years[year] = data.loc[(data['release_year'] == year) & (data['production_companies'].str.contains('Warner Bros'))]['income'].sum()
    
display(years)

years = pd.Series(years, index=years_list)
display(years.idxmax())

Counter({2014: 2295464519,
         2008: 2134595031,
         2012: 1258020056,
         2010: 1974712985,
         2015: 870368348})

2014

# 27. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. **Сентябрь**
5. Май

In [99]:
month_list = {1: 'Январь', 6: 'Июнь', 12: 'Декабрь', 9: 'Сентябрь', 5: 'Май'}

month_counter = Counter()
for month in month_list:
    month_counter[month] = len(data.loc[data['release_month'] == str(month)])
    
display(month_counter)

month_counter = pd.Series(month_counter, index=month_list)
print(month_list[month_counter.idxmax()])

Counter({1: 110, 6: 147, 12: 191, 9: 227, 5: 140})

Сентябрь


# 28. Сколько суммарно вышло фильмов летом? (за июнь, июль, август)
Варианты ответа:
1. 345
2. **450**
3. 478
4. 523
5. 381

In [108]:
month_list = ['6', '7', '8']

movies_count = data['release_month'].isin(month_list).sum()
print(movies_count)

450


# 29. Какой режисер выпускает (суммарно по годам) больше всего фильмов зимой?
Варианты ответов:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. **Peter Jackson**

In [102]:
directors_list = ['Steven Soderbergh', 'Clint Eastwood', 'Ridley Scott', 'Christopher Nolan', 'Peter Jackson']

directors = Counter()
for director in directors_list:
    directors[director] = len(data.loc[(data['director'].str.match(director)) & (data['release_month'].isin(['12', '1', '2']))])
    
display(directors)

directors = pd.Series(directors, index=directors_list)
display(directors.idxmax())

Counter({'Steven Soderbergh': 6,
         'Clint Eastwood': 6,
         'Ridley Scott': 3,
         'Christopher Nolan': 0,
         'Peter Jackson': 7})

'Peter Jackson'

# 30. Какой месяц чаще всего по годам самый прибыльный?
Почему нужно считать по годам, а не просто взять самый прибыльный месяц за весь период? Киноиндустрия активно развивается, прибыль компаний растет с каждым годом (например, в 2014 г. индустрия заработала в 2 раза больше, чем в 2004). Поэтому прибыльность более поздних месяцев будет значительно перевешивать предыдущие, что может привести к искажению выводов.

Варианты ответа:
1. Январь
2. **Июнь**
3. Декабрь
4. Сентябрь
5. Май

In [103]:
month_list = {'1': 'Январь', '6': 'Июнь', '12': 'Декабрь', '9': 'Сентябрь', '5': 'Май'}

# Sum income by month/year.
income_df = data.groupby(['release_year', 'release_month'])['income'].sum().reset_index()
# display(income_df)

# Filter dataframe by allowed months.
income_df = income_df.loc[(income_df['release_month'].isin(month_list))]
# display(income_df)

years_list = income_df['release_year'].unique()
# display(years_list)

month_counter = Counter()
for year in years_list:
    # Filter by year.
    yearly_income_df = income_df.loc[(income_df['release_year'] == year)]
    # Find best month.
    best_month_df = yearly_income_df.loc[yearly_income_df['income'] == yearly_income_df['income'].max()].reset_index()
    month = best_month_df.iloc[0]['release_month']
    if month not in month_counter:
        month_counter[month] = 0
    month_counter[month] += 1
    
best_month = max(month_counter)
print(month_list[best_month])

Июнь


# 31. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. **Four By Two Productions**

In [87]:
# Add the count of the characters of the movie title.
data['movie_title_chars_count'] = data['original_title'].apply(lambda title: len(title))
# display(data)

production_companies_list = ['Universal Pictures', 'Paramount Pictures', 'Jim Henson Company, The', 'Warner Bros', 'Four By Two Productions']

production_companies = Counter()
for production_company in production_companies_list:
    production_companies[production_company] = data.loc[(data['production_companies'].str.contains(production_company, regex=False))]['movie_title_chars_count'].mean()
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

Counter({'Universal Pictures': 14.731428571428571,
         'Paramount Pictures': 17.0327868852459,
         'Jim Henson Company, The': 59.0,
         'Warner Bros': 15.251366120218579,
         'Four By Two Productions': 83.0})

'Four By Two Productions'

# 32. Названия фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. **Four By Two Productions**

In [88]:
def split_title(title):
    words = title.split(' ')
    return len(words)

# Add the count of the words of the movie title.
data['movie_title_words_count'] = data['original_title'].apply(split_title)
# display(data)

production_companies_list = ['Universal Pictures', 'Paramount Pictures', 'Jim Henson Company, The', 'Warner Bros', 'Four By Two Productions']

production_companies = Counter()
for production_company in production_companies_list:
    production_companies[production_company] = data.loc[(data['production_companies'].str.contains(production_company, regex=False))]['movie_title_words_count'].mean()
    
display(production_companies)

production_companies = pd.Series(production_companies, index=production_companies_list)
display(production_companies.idxmax())

Counter({'Universal Pictures': 2.617142857142857,
         'Paramount Pictures': 2.8688524590163933,
         'Jim Henson Company, The': 10.0,
         'Warner Bros': 2.73224043715847,
         'Four By Two Productions': 12.0})

'Four By Two Productions'

# 33. Сколько разных слов используется в названиях фильмов (без учета регистра)?
Варианты ответа:
1. 6540
2. 1002
3. **2461**
4. 28304
5. 3432

In [89]:
def split_title(title):
    # What about .,:!? and etc?
    words = title.lower().replace('&', '').split(' ')
    return words

# Add the words extracted from the movie title.
data['movie_title_words'] = data['original_title'].apply(split_title)
# display(data)

words = data['movie_title_words'].explode().unique()
display(words)
print(len(words))

array(['jurassic', 'world', 'mad', ..., 'bullwinkle', 'hanging', 'crowd'],
      dtype=object)

2461


# 34. Какие фильмы входят в 1 процент лучших по рейтингу?
Варианты ответа:
1. **Inside Out, Gone Girl, 12 Years a Slave**
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King
4. 300, Lucky Number Slevin
5. Upside Down, 300, Inside Out, The Lord of the Rings: The Return of the King

In [107]:
s = data.sort_values(by='vote_average', ascending=False).loc[data['vote_average'] >= data['vote_average'].quantile(0.99)]
display(s.head(10))

imdb_id  popularity     budget     revenue  \
600   tt0468569    8.466668  185000000  1001921825   
125   tt2084970    8.110711   14000000   233555708   
9     tt2096673    6.326804  175000000   853708609   
118   tt0816692   24.949134  165000000   621752480   
34    tt3170832    3.557846    6000000    35401758   
1082  tt0167260    7.122455   94000000  1118888979   
873   tt0253474    2.364204   35000000   120072577   
1801  tt0209144    3.095625    9000000    39723096   
1192  tt2024544    3.625529   20000000   187000000   
128   tt2267998    6.438727   61000000   369330363   

                                     original_title  \
600                                 The Dark Knight   
125                              The Imitation Game   
9                                        Inside Out   
118                                    Interstellar   
34                                             Room   
1082  The Lord of the Rings: The Return of the King   
873                                     The Pianist   
1801                                        Memento   
1192                               12 Years a Slave   
128                                       Gone Girl   

                                                   cast           director  \
600   Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   
125   Benedict Cumberbatch|Keira Knightley|Matthew G...      Morten Tyldum   
9     Amy Poehler|Phyllis Smith|Richard Kind|Bill Ha...        Pete Docter   
118   Matthew McConaughey|Jessica Chastain|Anne Hath...  Christopher Nolan   
34    Brie Larson|Jacob Tremblay|Joan Allen|Sean Bri...   Lenny Abrahamson   
1082  Elijah Wood|Ian McKellen|Viggo Mortensen|Liv T...      Peter Jackson   
873   Adrien Brody|Thomas Kretschmann|Frank Finlay|M...     Roman Polanski   
1801  Guy Pearce|Carrie-Anne Moss|Joe Pantoliano|Mar...  Christopher Nolan   
1192  Chiwetel Ejiofor|Michael Fassbender|Lupita Nyo...      Steve McQueen   
128   Ben Affleck|Rosamund Pike|Carrie Coon|Neil Pat...      David Fincher   

                                                tagline  \
600                                     Why So Serious?   
125   The true enigma was the man who cracked the code.   
9              Meet the little voices inside your head.   
118   Mankind was born on Earth. It was never meant ...   
34                             Love knows no boundaries   
1082                    The eye of the enemy is moving.   
873   Music was his passion. Survival was his master...   
1801                  Some memories are best forgotten.   
1192    The extraordinary true story of Solomon Northup   
128         You don't know what you've got 'til it's...   

                                               overview  runtime  ...  \
600   Batman raises the stakes in his war on crime. ...      152  ...   
125   Based on the real life story of legendary cryp...      113  ...   
9     Growing up can be a bumpy road, and it's no ex...       94  ...   
118   Interstellar chronicles the adventures of a gr...      169  ...   
34    Jack is a young boy of 5 years old who has liv...      117  ...   
1082  Aragorn is revealed as the heir to the ancient...      201  ...   
873   The Pianist is a film adapted from the biograp...      150  ...   
1801  Suffering short-term memory loss after a head ...      113  ...   
1192  In the pre-Civil War United States, Solomon No...      134  ...   
128   With his wife's disappearance having become th...      145  ...   

                                   production_companies release_date  \
600   DC Comics|Legendary Pictures|Warner Bros.|Syncopy    7/16/2008   
125              Black Bear Pictures|Bristol Automotive   11/14/2014   
9     Walt Disney Pictures|Pixar Animation Studios|W...     6/9/2015   
118   Paramount Pictures|Legendary Pictures|Warner B...    11/5/2014   
34    Element Pictures|No Trace Camping|A24|Duperele...   10/16/2015   
1082                      WingNut Films|New Line Cinema    12/

# 35. Какие актеры чаще всего снимаются в одном фильме вместе?
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Hugh Jackman & Ian McKellen
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. **Daniel Radcliffe & Rupert Grint**

In [91]:
pairs = [
    ['Johnny Depp', 'Helena Bonham Carter'],
    ['Hugh Jackman', 'Ian McKellen'],
    ['Vin Diesel', 'Paul Walker'],
    ['Adam Sandler', 'Kevin James'],
    ['Daniel Radcliffe', 'Rupert Grint'],
]

artists_counter = Counter()
artists_list = []
for pair in pairs:
    artists_pair = ' & '.join(pair)
    artists_list.append(artists_pair)
    artists_counter[artists_pair] = len(data.loc[(data['cast'].str.contains(pair[0])) & (data['cast'].str.contains(pair[1]))])

display(artists_counter)  

artists_counter = pd.Series(artists_counter, index=artists_list)
display(artists_counter.idxmax())

Counter({'Johnny Depp & Helena Bonham Carter': 6,
         'Hugh Jackman & Ian McKellen': 5,
         'Vin Diesel & Paul Walker': 5,
         'Adam Sandler & Kevin James': 5,
         'Daniel Radcliffe & Rupert Grint': 8})

'Daniel Radcliffe & Rupert Grint'

# 36. У какого из режиссеров самый высокий процент фильмов со сборами выше бюджета?
Варианты ответа:
1. Quentin Tarantino
2. Steven Soderbergh
3. Robert Rodriguez
4. Christopher Nolan
5. Clint Eastwood

In [92]:
directors_list = ['Quentin Tarantino', 'Steven Soderbergh', 'Robert Rodriguez', 'Christopher Nolan', 'Clint Eastwood']

directors = Counter()
for director in directors_list:
    directors[director] = len(data.loc[(data['director'].str.match(director)) & (data['income'] > 0)])
    
display(directors)

directors = pd.Series(directors, index=directors_list)
display(directors.idxmax())

Counter({'Quentin Tarantino': 5,
         'Steven Soderbergh': 10,
         'Robert Rodriguez': 8,
         'Christopher Nolan': 8,
         'Clint Eastwood': 10})

'Steven Soderbergh'